includes/enrolment/class-sensei-enrolment-provider-journal-store.php

```diff
@@ -0,0 +1,283 @@
+
+/**
+ * This class is responsible for storing provider metadata like logs and status history.
+ */
+class Sensei_Enrolment_Provider_Journal_Store implements JsonSerializable {
+	const META_PREFIX_ENROLMENT_PROVIDERS_JOURNAL = 'sensei_enrolment_providers_journal_';
+
+	/**
+	 * Flag for if a state the store has changed.
+	 *
+	 * @var bool
+	 */
+	private $has_changed = false;
+
+	/**
+	 * Journal objects for the providers.
+	 *
+	 * @var Sensei_Enrolment_Provider_Journal[]
+	 */
+	private $providers_journal;
+
+	/**
+	 * User ID that this store is used for.
+	 *
+	 * @var int
+	 */
+	private $user_id;
+
+	/**
+	 * Course post ID that this store is used for.
+	 *
+	 * @var int
+	 */
+	private $course_id;
+
+	/**
+	 * Keeps track of instances of this class.
+	 *
+	 * @var self[][]
+	 */
+	private static $instances = [];
+
+	/**
+	 * Class constructor.
+	 *
+	 * @param int $user_id   User ID.
+	 * @param int $course_id Course post ID.
+	 */
+	private function __construct( $user_id, $course_id ) {
+		$this->user_id           = $user_id;
+		$this->course_id         = $course_id;
+		$this->providers_journal = [];
+	}
+
+	/**
+	 * Get a journal store record for a user/course.
+	 *
+	 * @param int $user_id   User ID.
+	 * @param int $course_id Course post ID.
+	 *
+	 * @return self
+	 */
+	private static function get( $user_id, $course_id ) {
+		if ( ! isset( self::$instances[ $user_id ] ) ) {
+			self::$instances[ $user_id ] = [];
+		}
+
+		if ( ! isset( self::$instances[ $user_id ][ $course_id ] ) ) {
+			self::$instances[ $user_id ][ $course_id ] = new self( $user_id, $course_id );
+
+			$provider_journal_stores = get_user_meta( $user_id, self::get_providers_meta_key( $course_id ), true );
+			if ( ! empty( $provider_journal_stores ) ) {
+				self::$instances[ $user_id ][ $course_id ]->restore_from_json( $provider_journal_stores );
+			}
+		}
+
+		return self::$instances[ $user_id ][ $course_id ];
+	}
+
+	/**
+	 * Restore a provider journal store from a serialized JSON string.
+	 *
+	 * @param string $json_string JSON representation of enrolment state.
+	 */
+	private function restore_from_json( $json_string ) {
+		$json_arr = json_decode( $json_string, true );
+
+		if ( ! $json_arr ) {
+			return;
+		}
+
+		foreach ( $json_arr as $provider_id => $provider_journal_data ) {
+			$provider_journal = Sensei_Enrolment_Provider_Journal::from_serialized_array( $provider_journal_data );
+
+			if ( $provider_journal ) {
+				$this->providers_journal[ $provider_id ] = $provider_journal;
+			}
+		}
+	}
+
+	/**
+	 * Return object that can be serialized by `json_encode()`.
+	 *
+	 * @return array
+	 */
+	public function jsonSerialize() {
+		return $this->providers_journal;
+	}
+
+	/**
+	 * Persist this store.
+	 *
+	 * @return bool
+	 */
+	public function save() {
+		/**
+		 * Enables journal storage for sensei.
+		 *
+		 * @since 3.0.0
+		 *
+		 * @param bool $enable_journal True to enable.
+		 */
+		if ( ! apply_filters( 'sensei_enable_journal', false ) ) {
```

Comment:
We should probably throw `enrolment` in here somewhere. `sensei_enable_enrolment_provider_journal`?

Reply:
Fixed! d661b2f542275a24e143b5cbf161ce48f8f3368c